### PR TITLE
fix(deps): update chokidar to v4

### LINKED ISSUE #5214: Chore: bump chokidar to v4
### Feature Request Checklist

- [x] I have read and agree to Mocha's [Code of Conduct](https://github.com/mochajs/mocha/blob/main/.github/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/mochajs/mocha/blob/main/.github/CONTRIBUTING.md)
- [x] I have searched for [related issues](https://github.com/mochajs/mocha/issues?q=is%3Aissue) and [issues with the `faq` label](https://github.com/mochajs/mocha/issues?utf8=%E2%9C%93&q=is%3Aissue%20label%3Afaq%20), but none matched my issue.
- [x] I want to provide a PR to resolve this

### Overview

I've made a PR to bump chokidar v4. It drops 12 dependencies and has better types.

### Suggested Solution

Bump chokidar to v4.

### Alternatives

Not bump chokidar to v4.

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by ggodlewski**
## PR Checklist

- [x] Addresses an existing open issue: fixes #5214
- [x] That issue was marked as [`status: accepting prs`](https://github.com/mochajs/mocha/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/mochajs/mocha/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

Current chokidar is bundled with tons of spam deps including `braces` which trigger some audit alert.

All test are working. It supports globs unlike https://github.com/mochajs/mocha/pull/5215

### CODE REVIEW

File: lib/cli/watch-run.js [hunk 1]
```diff
@@ -167,8 +168,36 @@ const createWatcher = (
   // we handle global fixtures manually
   mocha.enableGlobalSetup(false).enableGlobalTeardown(false);
 
-  const watcher = chokidar.watch(watchFiles, {
-    ignored: watchIgnore,
+  const watchFilesFlat = [];
```
Comment (JoshuaKGoldberg) [inline lib/cli/watch-run.js:171]: [Performance] This should be a `Set` so that lookups are an O(1) `.has()`, not an O(N) `.includes`. The `watchFilesFlat.includes(filePath)`s under `watcher.on('all', ...)` strike me as a likely hot path for repositories with many files.

File: lib/cli/watch-run.js [hunk 2]
```diff
@@ -184,8 +213,17 @@ const createWatcher = (
...
+  watcher.on('all', (event, filePath) => {
+    if (watchFilesFlat.includes(filePath)) {
+      rerunner.scheduleRun();
+    } else {
+      if (event === 'add') {
+        regenerateWatchFilesFlat();
+        if (watchFilesFlat.includes(filePath)) {
+          rerunner.scheduleRun();
+        }
+      }
```
Comment (JoshuaKGoldberg) [inline lib/cli/watch-run.js:225]: [Question] Is there a reason not to always regenerate if `event` is `'add`'? I.e.:

```js
watcher.on('all', (event, filePath) => {
  if (event === 'add') {
    regenerateWatchFilesFlat();
  }

  if (watchFilesFlat.includes(filePath)) {
    rerunner.scheduleRun();
  }
});
```

I think being able to deduplicate the code makes it a little clearer to read.

File: lib/cli/watch-run.js [hunk 3]
```diff
@@ -167,8 +168,36 @@ const createWatcher = (
...
+            }
+          }
+          return true;
+        }));
+    }
+    watchFilesFlat.splice(0, watchFiles.length);
+    watchFilesFlat.push(...newWatchFilesFlat);
+  }
+
+  regenerateWatchFilesFlat();
```
Comment (JoshuaKGoldberg) [inline lib/cli/watch-run.js:198]: [Design] It took me a good amount of reading to figure out what's going on here. Summarizing how I understand it: `watchFilesFlat` is the persistent reference used under the `watcher.all('all', ...)` event. `regenerateWatchFilesFlat` is called on startup and whenever a file is added to regenerate that list of files, factoring in file and ignore globs. This used to be done by Chokidar v3 but is no longer done in Chokidar v4.

Proposal: I think this file is pretty long, and this area of logic is pretty insular to itself. What do you think about extracting it to a separate class or function? Vaguely:

```diff
- const watchFilesFlat = [];
+ const globFilesTracker = new GlobFilesTracker(watchFiles, watchIgnore);

...

- regenerateWatchFilesFlat();
+ globFilesTracker.regenerate();
```

...where maybe `GlobFilesTracker` is in a separate file and contains the logic to regenerate?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Hey thanks for getting this started! Sorry it took so long to review - since Chokidar didn't end up needing to be bumped with v11, this fell by the wayside. But I'm excited to help push this forward along now! 🚀 

Requesting changes primarily on a performance point, as well as a design point. Happy to talk if you want. 🙂

**Comment by ggodlewski** [comment]
Hey @JoshuaKGoldberg, there is no way I remember the details after a month :). All I know it's been a struggle to get all tests working. In the meantime we switched to ava to resolve our issues, so I can't do much.
I updated the code to fix your issues. If there is anything that I can do fast let me know.

File: lib/cli/watch-run.js (hunk 1 above)
Comment (ggodlewski) [reply lib/cli/watch-run.js]: I learnt something. Thanks.

**Comment by talentlessguy** [comment]
@JoshuaKGoldberg is there any blocker to this PR besides merge conflicts of a lockfile?

**Comment by JoshuaKGoldberg** [comment]
Just a re-request for review. I'll add it to the queue for review :)

File: lib/cli/watch-run.js
Comment (JoshuaKGoldberg) [inline lib/cli/watch-run.js:1]: [Bug] I tried this PR out on real-world code and found a performance issue from files being renamed. I found it on the ESLint codebase but also reproduced it in isolation here: https://github.com/JoshuaKGoldberg/repros/tree/mocha-chokidar-v4-pr-watching. Could you please take a look?

tl;dr: in a minimal codebase with two root-level test files, re-watching is reported as taking >200ms

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Source code looks good, thanks for the refactor!

I finally found time to really dig in and try this in production - and think I found a performance bug. LMK what you think of it.

File: lib/cli/watch-run.js
Comment (ggodlewski) [reply lib/cli/watch-run.js:1]: How exactly are you renaming the file? I run `mv b.test.js z.test.js` and I don't have the state with 0 passing:

```
$ node ../mocha/bin/mocha.js "*.test.js" --watch
  a
    ✔ passes
  b
    ✔ passes
  2 passing (2ms)

ℹ [mocha] waiting for changes...

  a
    ✔ passes
  b
    ✔ passes

  2 passing (1ms)

ℹ [mocha] waiting for changes...
```

Comment (JoshuaKGoldberg) [reply lib/cli/watch-run.js:1]: Interesting. I did it through the VS Code UI. This is on an m1 Mac.

Comment (ggodlewski) [reply lib/cli/watch-run.js:1]: I'm running Ubuntu. I have no hardware to reproduce it. It's strange because the `tracker.regenerate();` is only triggered on `event === 'add'`. I don't see how the list of files can be empty during rename of 1 file. Maybe you can add some debug output so we can trace the tracker method calls?

Comment (JoshuaKGoldberg) [reply lib/cli/watch-run.js:1]: 👍 added to the README.md. Interestingly, the long delay of ~200ms only happens if an `npm install` populated `node_modules/`. If you `rm -rf node_modules/` then the delay is only around 20-30ms.

Comment (ggodlewski) [reply lib/cli/watch-run.js:1]: I added one more improvement. Glob should not traverse `node_modules`. Can you check if it improves anything?

Comment (JoshuaKGoldberg) [reply lib/cli/watch-run.js:1]: 💯 that fixed it for me! Back to 0-1ms  on the middle waiting.

+1 that it's a generalized solution too, not specific to `node_modules`. Nice 🙂

File: lib/cli/watch-run.js [hunk 4]
```diff
@@ -136,6 +137,43 @@ exports.watchRun = (mocha, {watchFiles, watchIgnore}, fileCollectParams) => {
...
+  regenerate() {
+    const watchIgnoreSet = new Set();
+    for (const pattern of this.watchIgnore) {
+      glob.sync(pattern, { dot: true }).forEach(filePath => watchIgnoreSet.add(filePath));
+    }
+
+    const globOpts = {
+      dot: true,
+      ignore: {
+        ignored: () => false,
```
Comment (JoshuaKGoldberg) [inline lib/cli/watch-run.js:156]: Despite the docs suggesting the object needs both `childrenIgnored` and `ignored`, I omitted `ignored` with no adverse effects locally. I think we can remove this?

```suggestion
```

**Review by JoshuaKGoldberg — APPROVED** [review]
🔥 This is fantastic, thanks a bunch @ggodlewski! I really appreciate you sticking through & working with me on the reviews.

I've played with this a bunch locally on relatively small and large repros, and haven't found any issues. I'd really prefer to get another review from someone on @mochajs/maintenance-crew just to be safe - but otherwise ✅ on my end.

Nice job!

**Comment by mark-wiemer** [comment]
I'll take a look at this tonight :)

**Comment by mark-wiemer** [comment]
@ggodlewski are you able to resolve the merge conflict? Otherwise looks great :)

**Review by mark-wiemer — APPROVED** [review]

**Comment by ggodlewski** [comment]
@mark-wiemer I rebased against main and updated lock file.

File: lib/cli/watch-run.js [hunk 5]
```diff
@@ -167,8 +204,10 @@ const createWatcher = (
   // we handle global fixtures manually
   mocha.enableGlobalSetup(false).enableGlobalTeardown(false);
 
-  const watcher = chokidar.watch(watchFiles, {
-    ignored: watchIgnore,
+  const tracker = new GlobFilesTracker(watchFiles, watchIgnore);
+  tracker.regenerate();
+
+  const watcher = chokidar.watch('.', {
```
Comment (mark-wiemer) [inline lib/cli/watch-run.js:210]: @ggodlewski if you're available, I'm debugging #5355 and curious why we changed this to `watch('.'` from `watch(watchFiles`

Comment (ggodlewski) [reply lib/cli/watch-run.js:210]: @mark-wiemer probably something related to removal of globs from v4. It has to start from some directory.
https://github.com/paulmillr/chokidar?tab=readme-ov-file#upgrading
I don't remember exactly, I'm guessing.

Comment (mark-wiemer) [reply lib/cli/watch-run.js:210]: You're right, thanks :D the author of the mentioned incident linked the same stuff. Thanks for your patience, I'm still getting my bearings here :)
